> Yes, I removed the caller and forgot to update the name.

Ok, well if I change ```fastParseASCIIIdentifierSSE42``` to ```fastParseASCIIIdentifier``` I still get
```
../../clang/lib/Lex/Lexer.cpp:1937:1: error: unused function 'fastParseASCIIIdentifier' [-Werror,-Wunused-function]
 1937 | fastParseASCIIIdentifier(const char *CurPtr, const char *BufferEnd) {
      | ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
so the optimized version still looks unused.

--- a/clang/lib/Lex/Lexer.cpp
+++ b/clang/lib/Lex/Lexer.cpp
@@ -1929,13 +1929,16 @@ static const char *fastParseASCIIIdentifierScalar(const char *CurPtr) {
 }
 
 // Fast path for lexing ASCII identifiers using SSE4.2 instructions.
-// Only enabled on x86/x86_64 when building with a compiler that supports
-// the 'target' attribute, which is used for runtime dispatch. Otherwise, we
-// fall back to the scalar implementation.
+// Only enabled on x86/x86_64 when building with __SSE4_2__ enabled, or with a
+// compiler that supports the 'target' attribute, used for runtime dispatch.
+// Otherwise, we fall back to the scalar implementation.
+// We avoid runtime check on Windows because it is not yet well-supported.
 #if defined(__SSE4_2__) || (defined(__i386__) || defined(__x86_64__)) &&       \
                                defined(__has_attribute) &&                     \
                                __has_attribute(target) && !defined(_WIN32)
-__attribute__((target("sse4.2"))) static const char *
+// "used" is a hack to suppress a false-positive warning due to a bug in
+// clang-18 and less. See PR175452.
+__attribute__((target("sse4.2"), used)) static const char *
 fastParseASCIIIdentifier(const char *CurPtr, const char *BufferEnd) {
   alignas(16) static constexpr char AsciiIdentifierRange[16] = {
       '_', '_', 'A', 'Z', 'a', 'z', '0', '9',
@@ -1965,9 +1968,7 @@ __attribute__((target("default")))
 #endif
 #endif
 #ifndef __SSE4_2__
-static const char *
-fastParseASCIIIdentifier(const char *CurPtr,
-                         [[maybe_unused]] const char *BufferEnd) {
+static const char *fastParseASCIIIdentifier(const char *CurPtr, const char *) {
   return fastParseASCIIIdentifierScalar(CurPtr);
 }
 #endif

--- a/clang/lib/Lex/Lexer.cpp
+++ b/clang/lib/Lex/Lexer.cpp
@@ -1935,7 +1935,7 @@ static const char *fastParseASCIIIdentifierScalar(const char *CurPtr) {
 #if (defined(__i386__) || defined(__x86_64__)) && defined(__has_attribute) &&  \
     __has_attribute(target) && !defined(_WIN32)
 __attribute__((target("sse4.2"))) static const char *
-fastParseASCIIIdentifierSSE42(const char *CurPtr, const char *BufferEnd) {
+fastParseASCIIIdentifier(const char *CurPtr, const char *BufferEnd) {
   alignas(16) static constexpr char AsciiIdentifierRange[16] = {
       '_', '_', 'A', 'Z', 'a', 'z', '0', '9',
   };

--- a/clang/lib/Lex/Lexer.cpp
+++ b/clang/lib/Lex/Lexer.cpp
@@ -1932,8 +1932,9 @@ static const char *fastParseASCIIIdentifierScalar(const char *CurPtr) {
 // Only enabled on x86/x86_64 when building with a compiler that supports
 // the 'target' attribute, which is used for runtime dispatch. Otherwise, we
 // fall back to the scalar implementation.
-#if (defined(__i386__) || defined(__x86_64__)) && defined(__has_attribute) &&  \
-    __has_attribute(target) && !defined(_WIN32)
+#if defined(__SSE4_2__) || (defined(__i386__) || defined(__x86_64__)) &&       \
+                               defined(__has_attribute) &&                     \
+                               __has_attribute(target) && !defined(_WIN32)
 __attribute__((target("sse4.2"))) static const char *
 fastParseASCIIIdentifier(const char *CurPtr, const char *BufferEnd) {
   alignas(16) static constexpr char AsciiIdentifierRange[16] = {
@@ -1959,13 +1960,17 @@ fastParseASCIIIdentifier(const char *CurPtr, const char *BufferEnd) {
   return fastParseASCIIIdentifierScalar(CurPtr);
 }
 
+#ifndef __SSE4_2__
 __attribute__((target("default")))
 #endif
+#endif
+#ifndef __SSE4_2__
 static const char *
 fastParseASCIIIdentifier(const char *CurPtr,
                          [[maybe_unused]] const char *BufferEnd) {
   return fastParseASCIIIdentifierScalar(CurPtr);
 }
+#endif
 
 bool Lexer::LexIdentifierContinue(Token &Result, const char *CurPtr) {
   // Match [_A-Za-z0-9]*, we have already matched an identifier start.

